keps/sig-apps/0753-sidecarcontainers.md

```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
+```yaml
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: worker
+        command: ["do a job"]
+      - name: "sidecar"
+        command: ["help"]
+  backoffLimit: 4
+  primaryContainer: worker
+```
 
-Another alternative would be to change the Job Spec to have a `primaryContainer` field to tell it which containers are important. However I feel this is perhaps too specific to job when this Sidecar concept could be useful in other scenarios.
+### Sidecar: true
+```yaml
+containers:
+  - name: myApp
+  - name: mySidecar
+    sidecar: true
+```
+A boolean flag of `sidecar: true` could be used to indicate which pods are sidecars, this was dismissed as it was considered too specific and potentially other types of container lifecycle may want to be added in the future.
 
-A boolean flag of `sidecar: true` could be used to indicate which pods are sidecars, however this prevents additional ContainerLifecycles from being added in the future.
+### terminationFatalToPod: true
+This suggestion was to have the ability to mark certain containers as critical to the pod, if they exited it would cause the other containers to exit. While this helped solve things like Jobs it didn't solve the wider issue of ordering startup and shutdown.
+
+```yaml
+containers:
+  - name: myApp
+    terminationFatalToPod: true
+  - name: mySidecar
+```
+
+### Depends On
+Similar to [systemd](https://www.freedesktop.org/wiki/Software/systemd/) this would allow you to specify that a container depends on another container, preventing that container from starting until the container it depends on has also started. This could also be used in shutdown to ensure that the containers which have dependent containers are only terminated after their dependents have all safely shut down.
+```yaml
+containers:
+  - name: myApp
+    dependsOn: mySidecar
+  - name: mySidecar
+```
+This was rejected as the UX was considered to be overly complicated for the use cases that we were trying to solve. It also doesn't solve the problem of Job termination where you want the sidecars to be terminated once the main containers have exited, to do that you would require some additional fields that would further complicate the UX.
+
+### Startup/Shutdown Phases/Numbers
+There were a few variations of this but they all had a similar idea which was the ability to order both the shutdown and startup of containers using phases or numbers to determine the ordering.
+examples:
+```yaml
+containers:
+  - name: myApp
+    StartupPhase: default
+  - name: mySidecar
+    StartupPhase: post-init
+```
+
+```yaml
+containers:
+  - name: myApp
+    startupOrder: 2
+    shutdownOrder: 1
+  - name: mySidecar
+    startupOrder: 2
+    shutdownOrder: 1
+  ```
+  This was dismissed as the UX was considered overly complex for the use cases we were trying to enable and also lacked the semantics around container shutdown triggering for things like Jobs.
```

Comment:
````suggestion
```
This was dismissed as the UX was considered overly complex for the use cases we were trying to enable and also lacked the semantics around container shutdown triggering for things like Jobs.
````